back/engines/commercial/multi_tenancy/lib/tasks/core/translate_project_to_all_locales.rake

```diff
@@ -0,0 +1,117 @@
+  task :translate_project_to_tenant_locales, %i[host project_id] => :environment do |_t, args|
+    Apartment::Tenant.switch(args[:host].tr('.', '_')) do
+      project = Project.find(args[:project_id])
+      locales = AppConfiguration.instance.settings['core']['locales']
+      translate_project = TranslateProjectMethods.new
+
+      puts "\nTranslating project: '#{project.title_multiloc['en']}' to all tenant locales\n\n"
+
+      translate_project.process_model(project, locales)
+
+      project.phases.each do |phase|
+        translate_project.process_model(phase, locales)
+
+        fields = phase&.custom_form&.custom_fields
+        fields&.each do |field|
+          translate_project.process_model(field, locales)
+          options = field&.options
+          options&.each { |option| translate_project.process_model(option, locales) }
+        end
+
+        report = phase&.report
+        layout = report&.layout
+        translate_project.process_layout(layout, locales) if layout
+        translate_project.process_data_units(report) if report
+      end
+
+      puts "\n--- end ---\n\n"
+    end
+  end
+end
+
+class TranslateProjectMethods
+  def process_model(model, locales)
+    multiloc_attributes = model.attributes.select { |attribute| attribute.ends_with? '_multiloc' }
+
+    multiloc_attributes.each do |multiloc_attribute|
+      attribute_name = multiloc_attribute[0]
+      translated = process_multiloc(multiloc_attribute[1], locales)
+      next unless translated
+
+      model[attribute_name] = translated
+    end
+
+    save_model(model)
+  end
+
+  def process_layout(layout, locales)
+    craftjs_json = layout.craftjs_json
+
+    craftjs_json.each do |k, v|
+      next unless v.is_a?(Hash)
+
+      text = v.dig('props', 'text')
+      next unless text.is_a?(Hash) && text['en'].present?
+
+      craftjs_json[k]['props']['text'] = process_multiloc(text, locales)
+    end
+
+    layout.craftjs_json = craftjs_json
+
+    save_model(layout)
+  end
+
+  def process_data_units(report)
+    user = User.find_by(email: 'moderator@citizenlab.co') || User.find_by(email: 'moderator@govocal.com')
+
+    if user.blank?
+      puts "No moderator user found, so could not re-publish graph_data_units for Report: #{report.id}"
+      nil
+    end
+
+    ReportBuilder::ReportPublisher.new(report, user).publish
+
+    puts "Re-published graph_data_units for Report: #{report.id}"
+  end
+
+  def process_multiloc(multiloc, locales)
+    english = multiloc['en']
+    return nil unless english
+
+    translator = MachineTranslations::MachineTranslationService.new
+
+    locales.each do |locale|
+      next if locale == 'en'
+
+      multiloc[locale] = translator.translate(english, 'en', locale)
+    end
+
+    multiloc
+  end
+
+  def save_model(model)
+    return unless model.changed?
+
+    if model.save
+      puts "Processed #{model.class} #{model.id}"
+    else
+      puts "Failed to save #{model.class} #{model.id}"
```

Comment:
It could be useful to keep the error and show it at the end of the execution. When errors appear higher up in the logs, it could be easy to miss them. There is also a `ScriptReporter` that could be used for this.

Reply:
Also a good idea, but I tend to just Ctrl-F and search for the `'Failed to save...'` string, so again not super important.